[Refactor]: Rename Script methods that only work on PreTapScript scripts

sanket1729 opened the pull request:
<s>
- This change is not really necessary as MAX_OPCODE check is only really
used while decoding transactions, but makes the code more consistent(and correct)

- The tests that use MAX_OPCODE are not failing because they are under P2SH context(where the OP_CHECKSIGADD is invalid anyways). The comment in the test has been updated to accurately reflect what is going on.</s>

Based on Aj's comment, I have changed the PR so that it renames the constants/functions that operate on PreTapScript Scripts. 

I don't know if this is the ideal way to separate pre-tapscripts and tapscripts.  If the new PR is not useful, feel free to close it.

prayank23 commented:
Concept ACK

Since BIP 342 is implemented in Bitcoin Core, OP_CHECKSIGADD is the last executable opcode now according to my understanding.

> This change is not really necessary as MAX_OPCODE check is only really
used while decoding transactions

Not sure about this part

sanket1729 commented:
> This change is not really necessary as MAX_OPCODE check is only really
used while decoding transactions

To elaborate, right now this is used while parsing scrpitPubkey and ScrpitSig in a Tx. It is impossible for `CHECKSIGADD` to end up in `scriptSig` or `scriptPubkey` in taproot. Secondly, the opcode execution is only valid for taproot scripts, but according to the current design, the functions checking validity of scripts don't have access to appropriate scriptContext.

So, the ideal solution would be to somehow pass the script context and enforce that the `CHECKSIGADD` is only used in tapscripts, but that probably requires more discussion and is out of scope for this PR.

benthecarman commented:
Concept ACK

theStack commented:
Concept ACK

JeremyRubin commented:
Would it be a better idea to split the baby now and make the handling of pre-v1 and v1 scripts different here?

laanwj commented on src/script/script.h:
```diff
@@ -205,7 +205,7 @@ enum opcodetype
 
 // Maximum value that an opcode can be
-static const unsigned int MAX_OPCODE = OP_NOP10;
+static const unsigned int MAX_OPCODE = OP_CHECKSIGADD;
```
context:
```c++
OP_NOP10 = 0xb9,
OP_CHECKSIGADD = 0xba,
```

laanwj commented:
Concept ACK

> So, the ideal solution would be to somehow pass the script context and enforce that the CHECKSIGADD is only used in tapscripts, but that probably requires more discussion and is out of scope for this PR.

The main thing to review here is then that this doesn't create any risk for non-tapscript verification. It seems that `MAX_OPCODE` is used in:
- `ParseOpCode` parsing from text (non-consensus critical)
- `CScript::HasValidOps()` (used only in `CheckTxScriptsSanity`, in `core_read`, and tests/fuzz, so non-consensus-critical)
- `ConsumeOpcodeType` in the fuzz tests (non-consensus critical)

I don't think this should even be labeled "consensus"? Or am I missing something.

Code review ACK 90f6a623b394e65a34a1918ec67420c899c52bf4

ajtowns commented:
I don't think this makes sense? For tapscript, the highest valid opcode is `OP_SUCCESS254`, and for non-tapscript the highest valid opcode is still `OP_NOP10`.

Maybe it would be better to rename `MAX_OPCODE` and `HasValidOps` to explicitly refer to "OG" script? (maybe also `ParseScript` in io_core.h)

maflcko reviewed (commented):
test fails? Maybe needs rebase?

```
test/script_parse_tests.cpp(53): error: in "script_parse_tests/parse_script": exception std::runtime_error expected but not raised
```

https://cirrus-ci.com/task/5542171511095296?logs=ci#L6208

fanquake commented:
@sanket1729 are you interested in following up here?

sanket1729 commented:
@fanquake, sorry I missed this. I will follow up on this in a couple of days.

sanket1729 commented:
> I don't think this makes sense? For tapscript, the highest valid opcode is OP_SUCCESS254, and for non-tapscript the highest valid opcode is still OP_NOP10.

Indeed

> Maybe it would be better to rename MAX_OPCODE and HasValidOps to explicitly refer to "OG" script? (maybe also ParseScript in io_core.h)

Agreed, this seems the cleaner way to solve this.

sanket1729 committed:
Rename script operations that only work on pretapscripts

Constant:
MAX_OPCODE -> MAX_OPCODE_PRE_TAPSCRIPT

Functions:
ParseScript -> ParseScriptPreTapScript
ParseOpcode -> ParseOpCodePreTapScript
HasValidOps -> HasValidOpsPreTapScript

class:
OpCodeParser -> PreTapScriptOpCodeParser

sanket1729 commented:
Changed the PR description. I think there might be other places where we implicitly assume script as PreTapScript. For example, there is also a constant `MAX_SCRIPT_SIZE` that only holds for Pretaproot scripts.

achow101 commented:
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.
